client/web/src/hooks/self-update.ts

```diff
@@ -0,0 +1,114 @@
+import { useEffect, useState } from "react"
+import { apiFetch } from "src/api"
+
+// see tailcfg.ClientVersion
+export type ClientVersion = {
+  RunningLatest: boolean
+  LatestVersion?: string
+}
+
+// see ipnstate.UpdateProgress
+export type UpdateProgress = {
+  status: "UpdateFinished" | "UpdateInProgress" | "UpdateFailed"
+  message: string
+  version: string
+}
+
+export enum UpdateState {
+  UpToDate,
+  Available,
+  InProgress,
+  Complete,
+  Failed,
+}
+
+export function useInstallUpdate(currentVersion: string, cv: ClientVersion) {
+  const [updateState, setUpdateState] = useState<UpdateState>(
+    cv.RunningLatest ? UpdateState.UpToDate : UpdateState.Available
+  )
+
+  const [updateLog, setUpdateLog] = useState<string>("")
+
+  const appendUpdateLog = (msg: string) => {
+    setUpdateLog(updateLog + msg + "\n")
+  }
+
+  useEffect(() => {
+    if (updateState !== UpdateState.Available) {
+      // useEffect cleanup function
+      return () => {}
+    }
+    apiFetch("/update", "POST").catch((err) => {
+      console.log(err)
+      setUpdateState(UpdateState.Failed)
+    })
+
+    setUpdateState(UpdateState.InProgress)
+
+    let tsAwayForPolls = 0
+    let updateMessagesRead = 0
+
+    let timer = 0
+
+    function poll() {
+      apiFetch("/update/progress", "GET")
+        .then((res) => res.json())
+        .then((res: UpdateProgress[]) => {
+          for (; updateMessagesRead < res.length; ++updateMessagesRead) {
+            const up = res[updateMessagesRead]
+            if (up.status === "UpdateFailed") {
+              setUpdateState(UpdateState.Failed)
+              if (up.message) appendUpdateLog("ERROR: " + up.message)
+              return
+            }
+
+            if (up.status === "UpdateFinished") {
+              // if update finished and tailscaled did not go away (ie. did not restart),
+              // then the version being the same might not be an error, it might just require
+              // the user to restart Tailscale manually (this is required in some cases in the
+              // clientupdate package).
+              if (up.version === currentVersion && tsAwayForPolls > 0) {
+                setUpdateState(UpdateState.Failed)
+                appendUpdateLog(
+                  "ERROR: Update failed, still running Tailscale " + up.version
+                )
+                if (up.message) appendUpdateLog("ERROR: " + up.message)
+              } else {
+                setUpdateState(UpdateState.Complete)
+                if (up.message) appendUpdateLog("INFO: " + up.message)
+              }
+              return
+            }
+
+            setUpdateState(UpdateState.InProgress)
+            if (up.message) appendUpdateLog("INFO: " + up.message)
+          }
+
+          timer = setTimeout(poll, 1000)
```

Comment:
Do we always want to poll again for all other responses?

Reply:
the only response that hasn't been handled at this point is `UpdateInProgress`, and we do want to poll again in that case

Reply:
What if they happened to update it themselves on the side? Not possible to get UpToDate here?

Reply:
this is a bit confusing, so I'm open to redoing this differently, but there's two "states" to keep track of - the state of the UI, which this timer is not dependent on (and which has an `UpToDate` state), and the state of the self-update coming in from the localapi endpoint, which is what the timer depends on, and which only has `UpdateFinished`, `UpdateInProgress`, and `UpdateFailed`

Reply:
Oooh okay yes I see the typing above just has those three states. Okay I'd throw a little comment in here that the state is UpdateInProgress

If we want to be extra safe, you could still check `if (up.status === "UpdateInProgress") {`, that way the addition of a new update value wouldn't break things in the future, but that's probably highly unlikely

Reply:
that won't be possible, since `up` is from the for loop scope :sweat_smile: I'll add the comment